File: docker_installer/installers/macos.py
"""macOS-specific Docker installer using Docker Desktop DMG."""

import os
import time
import tempfile
import subprocess
from typing import Optional
from .base import BaseInstaller, InstallOptions, InstallResult
from ..utils.system import SystemInfo, Architecture
from ..utils.cli import CLI
from ..utils.runner import CommandRunner


class MacOSInstaller(BaseInstaller):
    """Installer for Docker on macOS using Docker Desktop DMG."""

    # Docker Desktop download URLs
    DOCKER_DMG_URL_ARM64 = "https://desktop.docker.com/mac/main/arm64/Docker.dmg"
    DOCKER_DMG_URL_AMD64 = "https://desktop.docker.com/mac/main/amd64/Docker.dmg"

    def __init__(self, system_info: SystemInfo, cli: CLI, runner: CommandRunner):
        """Initialize the macOS Docker installer."""
        super().__init__(system_info, cli, runner)

    def _get_docker_dmg_url(self) -> str:
        """Get the appropriate Docker DMG URL for the current architecture."""
        if self.system_info.architecture == Architecture.ARM64:
            return self.DOCKER_DMG_URL_ARM64
        return self.DOCKER_DMG_URL_AMD64

    def check_existing_installation(self) -> bool:
        """Check if Docker is already installed."""
        # Check for docker command
        if self.runner.check_command_exists("docker"):
            # Verify docker daemon is accessible
            result = self.runner.run(
                ["docker", "info"],
                sudo=False,
                timeout=10
            )
            if result.success:
                return True

        # Check if Docker Desktop app exists
        docker_app_paths = [
            "/Applications/Docker.app",
            os.path.expanduser("~/Applications/Docker.app")
        ]

        for path in docker_app_paths:
            if os.path.exists(path):
                return True

        return False

    def install_docker(self) -> bool:
        """Install Docker Desktop by downloading and installing the official DMG."""
        dmg_url = self._get_docker_dmg_url()
        arch_name = "Apple Silicon" if self.system_info.architecture == Architecture.ARM64 else "Intel"

        self.cli.print_info(f"Telechargement de Docker Desktop pour {arch_name}...")
        self.cli.print_info(f"URL: {dmg_url}")

        # Create temporary directory for download
        with tempfile.TemporaryDirectory() as tmp_dir:
            dmg_path = os.path.join(tmp_dir, "Docker.dmg")

            # Download the DMG
            self.cli.print_progress("Telechargement en cours...")
            result = self.runner.run(
                ["curl", "-fSL", "-o", dmg_path, dmg_url],
                description="Telechargement de Docker Desktop",
                sudo=False,
                timeout=600  # 10 minutes for download
            )

            if not result.success:
                self.cli.print_error("Echec du telechargement de Docker Desktop")
                self.cli.print_info("Vous pouvez telecharger manuellement depuis:")
                self.cli.print_info("https://www.docker.com/products/docker-desktop/")
                return False

            self.cli.print_success("Telechargement termine")

            # Mount the DMG
            self.cli.print_info("Montage du DMG...")
            mount_point = "/Volumes/Docker"

            # Unmount if already mounted
            if os.path.exists(mount_point):
                self.runner.run(["hdiutil", "detach", mount_point, "-quiet"], sudo=False)

            result = self.runner.run(
                ["hdiutil", "attach", dmg_path, "-nobrowse", "-quiet"],
                description="Montage du DMG",
                sudo=False,
                timeout=60
            )

            if not result.success:
                self.cli.print_error("Echec du montage du DMG")
                return False

            try:
                # Copy Docker.app to /Applications
                self.cli.print_info("Installation de Docker Desktop dans /Applications...")
                docker_app_src = "/Volumes/Docker/Docker.app"

                if not os.path.exists(docker_app_src):
                    self.cli.print_error("Docker.app non trouve dans le DMG")
                    return False

                # Remove existing installation if present
                docker_app_dest = "/Applications/Docker.app"
                if os.path.exists(docker_app_dest):
                    self.cli.print_info("Suppression de l'ancienne version...")
                    result = self.runner.run(
                        ["rm", "-rf", docker_app_dest],
                        sudo=False
                    )
                    if not result.success:
                        self.cli.print_error("Impossible de supprimer l'ancienne version")
                        return False

                # Copy the app
                result = self.runner.run(
                    ["cp", "-R", docker_app_src, docker_app_dest],
                    description="Copie de Docker.app",
                    sudo=False,
                    timeout=120
                )

                if not result.success:
                    self.cli.print_error("Echec de la copie de Docker.app")
                    return False

                self.cli.print_success("Docker Desktop installe dans /Applications")
                return True

            finally:
                # Always unmount the DMG
                self.cli.print_info("Demontage du DMG...")
                self.runner.run(
                    ["hdiutil", "detach", mount_point, "-quiet"],
                    sudo=False,
                    timeout=30
                )

    def install_docker_compose(self) -> bool:
        """Docker Compose is included with Docker Desktop."""
        self.cli.print_info("Docker Compose est inclus avec Docker Desktop")
        return True

    def configure_docker(self) -> bool:
        """Configure Docker on macOS."""
        # Docker Desktop handles most configuration automatically
        self.cli.print_info("Docker Desktop gere la configuration automatiquement")

        # Add docker CLI to PATH if needed
        docker_cli_path = "/usr/local/bin/docker"
        if not os.path.exists(docker_cli_path):
            # Docker Desktop symlinks are usually created when the app starts
            self.cli.print_info("Les liens symboliques seront crees au demarrage de Docker Desktop")

        return True

    def start_docker(self) -> bool:
        """Start Docker Desktop application."""
        # Check if Docker is already running
        result = self.runner.run(
            ["docker", "info"],
            sudo=False,
            timeout=10
        )

        if result.success:
            self.cli.print_success("Docker est deja en cours d'execution")
            return True

        # Start Docker Desktop
        self.cli.print_info("Demarrage de Docker Desktop...")

        result = self.runner.run(
            ["open", "-a", "Docker"],
            description="Demarrage de Docker Desktop",
            sudo=False
        )

        if not result.success:
            self.cli.print_warning("Impossible de demarrer Docker Desktop automatiquement")
            self.cli.print_info("Veuillez demarrer Docker Desktop manuellement depuis Applications")
            return False

        # Wait for Docker to be ready
        self.cli.print_info("Attente du demarrage de Docker...")
        max_attempts = 30
        for i in range(max_attempts):
            time.sleep(2)
            result = self.runner.run(
                ["docker", "info"],
                sudo=False,
                timeout=5
            )
            if result.success:
                self.cli.print_success("Docker est pret!")
                return True

            if i < max_attempts - 1:
                self.cli.print_progress(f"Attente... ({i + 1}/{max_attempts})")

        self.cli.print_warning("Docker prend du temps a demarrer")
        self.cli.print_info("Veuillez attendre que l'icone Docker dans la barre de menu indique 'Running'")
        return False

    def install(self, options: InstallOptions) -> InstallResult:
        """Run the complete Docker installation for macOS."""
        result = super().install(options)

        if result.success:
            # Additional macOS-specific info
            self.cli.print_section("Informations supplementaires")
            self.cli.print_info("Docker Desktop inclut:")
            self.cli.print("  - Docker Engine")
            self.cli.print("  - Docker CLI")
            self.cli.print("  - Docker Compose")
            self.cli.print("  - Docker BuildKit")
            self.cli.print("  - Kubernetes (optionnel)")
            self.cli.print()
            self.cli.print_info("Pour activer Kubernetes:")
            self.cli.print("  Docker Desktop > Preferences > Kubernetes > Enable Kubernetes")

        return result

File: internal/installers/docker/macos.go
package docker

import (
	"os"
	"path/filepath"
	"time"

	"github.com/keyxmare/DevBootstrap/internal/installers"
	"github.com/keyxmare/DevBootstrap/internal/runner"
	"github.com/keyxmare/DevBootstrap/internal/system"
)

// Docker Desktop download URLs
const (
	dockerDMGURLARM64 = "https://desktop.docker.com/mac/main/arm64/Docker.dmg"
	dockerDMGURLAMD64 = "https://desktop.docker.com/mac/main/amd64/Docker.dmg"
)

// macOSInstaller handles Docker installation on macOS.
type macOSInstaller struct {
	*installers.BaseInstaller
}

// newMacOSInstaller creates a new macOS Docker installer.
func newMacOSInstaller(base *installers.BaseInstaller) *macOSInstaller {
	return &macOSInstaller{BaseInstaller: base}
}

// getDMGURL returns the appropriate DMG URL for the current architecture.
func (m *macOSInstaller) getDMGURL() string {
	if m.SystemInfo.Arch == system.ArchARM64 {
		return dockerDMGURLARM64
	}
	return dockerDMGURLAMD64
}

// CheckExisting checks if Docker is already installed.
func (m *macOSInstaller) CheckExisting() (installers.AppStatus, string) {
	// Check if docker command exists and daemon is accessible
	if m.Runner.CommandExists("docker") {
		result := m.Runner.Run([]string{"docker", "info"}, runner.WithTimeout(10*time.Second))
		if result.Success {
			version := m.Runner.GetCommandVersion("docker", "--version")
			return installers.StatusInstalled, version
		}
	}

	// Check if Docker Desktop app exists
	dockerAppPaths := []string{
		"/Applications/Docker.app",
		filepath.Join(m.SystemInfo.HomeDir, "Applications/Docker.app"),
	}

	for _, path := range dockerAppPaths {
		if _, err := os.Stat(path); err == nil {
			return installers.StatusInstalled, "(Docker Desktop installe)"
		}
	}

	return installers.StatusNotInstalled, ""
}

// Install installs Docker Desktop by downloading and installing the official DMG.
func (m *macOSInstaller) Install(opts *installers.InstallOptions) *installers.InstallResult {
	dmgURL := m.getDMGURL()
	archName := "Intel"
	if m.SystemInfo.Arch == system.ArchARM64 {
		archName = "Apple Silicon"
	}

	m.CLI.PrintInfo("Telechargement de Docker Desktop pour " + archName + "...")
	m.CLI.PrintInfo("URL: " + dmgURL)

	// Create temporary directory for download
	tmpDir, err := os.MkdirTemp("", "docker-install-*")
	if err != nil {
		return installers.NewFailureResult("Impossible de creer le repertoire temporaire", err.Error())
	}
	defer os.RemoveAll(tmpDir)

	dmgPath := filepath.Join(tmpDir, "Docker.dmg")

	// Download the DMG
	if !m.Runner.DownloadFile(dmgURL, dmgPath, runner.WithDescription("Telechargement de Docker Desktop")) {
		return installers.NewFailureResult(
			"Echec du telechargement de Docker Desktop",
			"Vous pouvez telecharger manuellement depuis: https://www.docker.com/products/docker-desktop/",
		)
	}
	m.CLI.PrintSuccess("Telechargement termine")

	// Mount the DMG
	m.CLI.PrintInfo("Montage du DMG...")
	mountPoint := "/Volumes/Docker"

	// Unmount if already mounted
	if _, err := os.Stat(mountPoint); err == nil {
		m.Runner.Run([]string{"hdiutil", "detach", mountPoint, "-quiet"})
	}

	result := m.Runner.Run(
		[]string{"hdiutil", "attach", dmgPath, "-nobrowse", "-quiet"},
		runner.WithDescription("Montage du DMG"),
		runner.WithTimeout(60*time.Second),
	)

	if !result.Success {
		return installers.NewFailureResult("Echec du montage du DMG", result.Stderr)
	}

	// Ensure we unmount at the end
	defer func() {
		m.CLI.PrintInfo("Demontage du DMG...")
		m.Runner.Run([]string{"hdiutil", "detach", mountPoint, "-quiet"}, runner.WithTimeout(30*time.Second))
	}()

	// Copy Docker.app to /Applications
	m.CLI.PrintInfo("Installation de Docker Desktop dans /Applications...")
	dockerAppSrc := "/Volumes/Docker/Docker.app"

	if _, err := os.Stat(dockerAppSrc); os.IsNotExist(err) {
		return installers.NewFailureResult("Docker.app non trouve dans le DMG")
	}

	// Remove existing installation if present
	dockerAppDest := "/Applications/Docker.app"
	if _, err := os.Stat(dockerAppDest); err == nil {
		m.CLI.PrintInfo("Suppression de l'ancienne version...")
		if err := m.Runner.RemoveAll(dockerAppDest); err != nil {
			return installers.NewFailureResult("Impossible de supprimer l'ancienne version", err.Error())
		}
	}

	// Copy the app
	copyResult := m.Runner.Run(
		[]string{"cp", "-R", dockerAppSrc, dockerAppDest},
		runner.WithDescription("Copie de Docker.app"),
		runner.WithTimeout(120*time.Second),
	)

	if !copyResult.Success {
		return installers.NewFailureResult("Echec de la copie de Docker.app", copyResult.Stderr)
	}

	m.CLI.PrintSuccess("Docker Desktop installe dans /Applications")

	// Start Docker Desktop
	m.startDocker()

	// Show additional info
	m.showFinalInfo()

	return installers.NewSuccessResult("Docker Desktop installe avec succes")
}

// startDocker starts Docker Desktop.
func (m *macOSInstaller) startDocker() {
	// Check if Docker is already running
	result := m.Runner.Run([]string{"docker", "info"}, runner.WithTimeout(10*time.Second))
	if result.Success {
		m.CLI.PrintSuccess("Docker est deja en cours d'execution")
		return
	}

	// Start Docker Desktop
	m.CLI.PrintInfo("Demarrage de Docker Desktop...")
	startResult := m.Runner.Run(
		[]string{"open", "-a", "Docker"},
		runner.WithDescription("Demarrage de Docker Desktop"),
	)

	if !startResult.Success {
		m.CLI.PrintWarning("Impossible de demarrer Docker Desktop automatiquement")
		m.CLI.PrintInfo("Veuillez demarrer Docker Desktop manuellement depuis Applications")
		return
	}

	// Wait for Docker to be ready
	m.CLI.PrintInfo("Attente du demarrage de Docker...")
	maxAttempts := 30
	for i := 0; i < maxAttempts; i++ {
		time.Sleep(2 * time.Second)
		result := m.Runner.Run([]string{"docker", "info"}, runner.WithTimeout(5*time.Second))
		if result.Success {
			m.CLI.PrintSuccess("Docker est pret!")
			return
		}
		m.CLI.PrintProgress("Attente... (" + string(rune('0'+i+1)) + "/" + string(rune('0'+maxAttempts)) + ")")
	}

	m.CLI.PrintWarning("Docker prend du temps a demarrer")
	m.CLI.PrintInfo("Veuillez attendre que l'icone Docker dans la barre de menu indique 'Running'")
}

// showFinalInfo shows additional information after installation.
func (m *macOSInstaller) showFinalInfo() {
	m.CLI.PrintSection("Informations supplementaires")
	m.CLI.PrintInfo("Docker Desktop inclut:")
	m.CLI.Println("  - Docker Engine")
	m.CLI.Println("  - Docker CLI")
	m.CLI.Println("  - Docker Compose")
	m.CLI.Println("  - Docker BuildKit")
	m.CLI.Println("  - Kubernetes (optionnel)")
	m.CLI.Println("")
	m.CLI.PrintInfo("Pour activer Kubernetes:")
	m.CLI.Println("  Docker Desktop > Preferences > Kubernetes > Enable Kubernetes")
}

// Verify verifies the Docker installation.
func (m *macOSInstaller) Verify() bool {
	result := m.Runner.Run([]string{"docker", "--version"}, runner.WithTimeout(10*time.Second))
	return result.Success
}

// Uninstall removes Docker Desktop.
func (m *macOSInstaller) Uninstall(opts *installers.UninstallOptions) *installers.UninstallResult {
	m.CLI.PrintSection("Desinstallation de Docker Desktop")

	// Quit Docker Desktop
	m.CLI.PrintInfo("Fermeture de Docker Desktop...")
	m.Runner.Run([]string{"osascript", "-e", `quit app "Docker"`})
	time.Sleep(2 * time.Second)

	// Remove Docker.app
	dockerApp := "/Applications/Docker.app"
	if _, err := os.Stat(dockerApp); err == nil {
		m.CLI.PrintInfo("Suppression de Docker.app...")
		if err := m.Runner.RemoveAll(dockerApp); err != nil {
			return installers.NewUninstallFailureResult("Impossible de supprimer Docker.app", err.Error())
		}
		m.CLI.PrintSuccess("Docker.app supprime")
	}

	// Remove Docker data directories
	if opts.RemoveData {
		dataDirectories := []string{
			filepath.Join(m.SystemInfo.HomeDir, ".docker"),
			filepath.Join(m.SystemInfo.HomeDir, "Library/Group Containers/group.com.docker"),
			filepath.Join(m.SystemInfo.HomeDir, "Library/Containers/com.docker.docker"),
			filepath.Join(m.SystemInfo.HomeDir, "Library/Application Support/Docker Desktop"),
			filepath.Join(m.SystemInfo.HomeDir, "Library/Preferences/com.docker.docker.plist"),
			filepath.Join(m.SystemInfo.HomeDir, "Library/Saved Application State/com.electron.docker-frontend.savedState"),
			filepath.Join(m.SystemInfo.HomeDir, "Library/Logs/Docker Desktop"),
		}

		m.CLI.PrintInfo("Suppression des donnees Docker...")
		for _, dir := range dataDirectories {
			if _, err := os.Stat(dir); err == nil {
				m.Runner.RemoveAll(dir)
			}
		}
		m.CLI.PrintSuccess("Donnees Docker supprimees")
	}

	return installers.NewUninstallSuccessResult("Docker Desktop desinstalle avec succes")
}
